### PR TITLE
Register memory info command on vmservice for Android devices

### LINKED ISSUE #41225: Register additional Service Extensions to Flutter Tool for DevTools Profiling
## Use case

Flutter DevTools today reports the resident set size (RSS) of a Flutter application.  The RSS is an opaque value (just the total size) not very informative.  We would like to display the RSS but we'd like to display additional information (from dumpsys meminfo) e.g., Java Heap, Native Heap, Code, Stack, Graphics, Private Other, and System that comprise the RSS.

To do this we would need to send adb shell commands e.g.,

`     adb shell dumpsys meminfo <package_name>`

Today we can't be connected to both the VMService and adb on the same Flutter app profile run.

If the last connection is DevTools then 'adb shell' causes DevTools to exit.

## Proposal

After talking to @jonahwilliams he believes this could be implemented by having the flutter tool register additional service extensions which ultimately shelled out to adb or other system tools. That would make it usable via anything that has a vmservice connection I think?

The caveat here is that the way our vmservice connections are currently designed, by the time we get to the vmservice connection we've forgotten what device we're running on. I think that would be fixable via some plumbing though.  I would also need to find the package but I could use:

`dumpsys window windows | grep -E 'mFocusedApp' `

This feature would allow DevTools to keep our VM Service connection (talking to the VM) as well as issuing adb shell commands dumpsys meminfo and parsing the stdout (as well as other interesting stuff like batterystats, netstats, etc.)

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Allows the Device sub-classes to provide a `queryRss` method which is connected to the vmservice. implements this on AndroidDevice by calling https://github.com/flutter/flutter/issues/41225 and parsing the results.

Also cleans up flutter device to support a single vmservice instance.

Fixes https://github.com/flutter/flutter/issues/41225

cc @terrylucas to determine what specific fields we need.

### CODE REVIEW

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 1]
```diff
@@ -353,6 +354,19 @@ class VMService {
     await _peer.done;
   }
 
+  /// The connected physical device instance.
+  Device _device;
+  set device(Device device) {
+    _device = device;
+    _peer.registerMethod('flutterRss', (rpc.Parameters params) async {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/vmservice.dart]: TBD: good method name

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 2]
```diff
@@ -188,8 +188,9 @@ class FlutterDevice {
       }
       printTrace('Successfully connected to service protocol: $observatoryUri');
 
-      vmServices = <VMService>[service];
-      device.getLogReader(app: package).connectedVMServices = vmServices;
+      vmService = service;
+      service.device = device;
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/resident_runner.dart]: This is where the vmservice learns about the device type.

**Comment by jonahwilliams** [comment]
FYI, @zanderso @jmagman

File: packages/flutter_tools/lib/src/android/android_device.dart [hunk 3]
```diff
@@ -713,6 +731,45 @@ Map<String, String> parseAdbDeviceProperties(String str) {
   return properties;
 }
 
+
+const String _kJavaHeapKey = 'Java Heap';
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/android/android_device.dart]: The string 'Java Heap' isn't in the example below or in the linked docs.

File: packages/flutter_tools/lib/src/android/android_device.dart [hunk 4]
```diff
@@ -713,6 +731,45 @@ Map<String, String> parseAdbDeviceProperties(String str) {
...
+/// ...
+///
+/// For more information, see https://developer.android.com/studio/command-line/dumpsys.
+@visibleForTesting
+Map<String, Object> parseMeminfoDump(String input) {
+  final Map<String, Object> result = <String, Object>{};
+  for (String line in input.split('\n')) {
+    line = line.trim();
+    if (line.startsWith(_kJavaHeapKey)) {
+      result[_kJavaHeapKey] = int.tryParse(line.split(':').last.trim());
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/android/android_device.dart]: There are no colons in the example above or in the linked documentation.

File: packages/flutter_tools/test/general.shard/android/android_device_test.dart [hunk 5]
```diff
@@ -637,6 +637,67 @@ flutter:
...
+    .ttf mmap       35        0       32        0
+    .dex mmap     2205        4     1172        0
+    .oat mmap       64        0        0        0
+    .art mmap     4228     3848       24        2
+   Other mmap    20713        4    20704        0
+    GL mtrack     2380     2380        0        0
+      Unknown    43971    43968        0        1
+        TOTAL    95085    61120    31392       62    23108    13495     9612
+
+ App Summary
```
Comment (zanderso) [inline packages/flutter_tools/test/general.shard/android/android_device_test.dart]: This section is missing from the documentation. Do different Android versions output different info?

File: packages/flutter_tools/lib/src/android/android_device.dart [hunk 6]
```diff
@@ -648,6 +651,21 @@ class AndroidDevice extends Device {
         (int exitCode) => exitCode == 0 || allowHeapCorruptionOnWindows(exitCode));
   }
 
+  @override
+  Future<Map<String, Object>> queryRss() async {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/android/android_device.dart]: Instead of returning a `Map`, consider promoting the return type here to a class with platform agnostic fields from a base class and android-specific fields from an inherited class.

Then, instead of naming this method `queryRss` you can name it `queryBaseClass` since it's really doing a lot more than just grabbing Rss.

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 7]
```diff
@@ -132,7 +132,7 @@ class FlutterDevice {
   final Device device;
   final ResidentCompiler generator;
   Stream<Uri> observatoryUris;
-  List<VMService> vmServices;
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/resident_runner.dart]: Maybe this cleanup should go in a separate PR.

File: packages/flutter_tools/lib/src/resident_runner.dart (hunk 2 above)
Comment (zanderso) [reply packages/flutter_tools/lib/src/resident_runner.dart]: Could you pass the device to `VMService.connect` ?

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 8]
```diff
@@ -353,6 +354,19 @@ class VMService {
     await _peer.done;
   }
 
+  /// The connected physical device instance.
+  Device _device;
+  set device(Device device) {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/vmservice.dart]: If you pass the device to VMService.connect, then maybe you don't need this complicated setter, and you can set up the extension in the constructor with all the other extensions.

File: packages/flutter_tools/lib/src/android/android_device.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/android/android_device.dart]: Updated to full example

File: packages/flutter_tools/test/general.shard/android/android_device_test.dart (hunk 5 above)
Comment (jonahwilliams) [reply packages/flutter_tools/test/general.shard/android/android_device_test.dart]: Added to documentation. Not that I know of

File: packages/flutter_tools/lib/src/resident_runner.dart (hunk 7 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/resident_runner.dart]: Removed

File: packages/flutter_tools/lib/src/vmservice.dart (hunk 8 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/vmservice.dart]: Done

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 9]
```diff
@@ -88,7 +88,9 @@ class FlutterDevice {
         fileSystemScheme: fileSystemScheme,
         targetModel: TargetModel.dartdevc,
         experimentalFlags: experimentalFlags,
-        platformDill: artifacts.getArtifactPath(Artifact.webPlatformKernelDill, mode: buildMode),
+        platformDill: fs.file(artifacts
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/resident_runner.dart]: This is an unrelated change, will revert

File: packages/flutter_tools/lib/src/android/android_device.dart (hunk 6 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/android/android_device.dart]: I would be converting that back to JSON almost immediately though, so it would be a fairly short-lived object.

I'll wait for terry to chime in with what information we actually need from this dumpinfo first.

Comment (zanderso) [reply packages/flutter_tools/lib/src/android/android_device.dart]: When adding support for this to a new platform, it will be good to have support from the language/type-system ensuring that the expected info is filled in.

Comment (terrylucas) [reply packages/flutter_tools/lib/src/android/android_device.dart]: Jonah I was hopping to get all information in the App Summary section:

Java Heap:
Native Heap:
Code::
Stack:
Graphics:
Private Other:
System:
TOTAL:

these numbers seem the most useful for Android users (or at least its what Android Studio reports in their memory profiler).

At some point, we we'll probably hookup to dumpsys battery and dumpsys network stats too.  These might be passed as a row(s) or data instead of lots of parsing (but that's in the future).  The JSON/Map seems fine for me as it will be coming via a service call.

File: packages/flutter_tools/lib/src/android/android_device.dart (hunk 4 above)
Comment (terrylucas) [reply packages/flutter_tools/lib/src/android/android_device.dart]: Parsing of 'TOTAL:' would be a little different because of 'TOTAL SWAP PSS:' on the same line I would ignore 'TOTAL:' as the other 7 will add up to this anyway.

File: packages/flutter_tools/test/general.shard/android/android_device_test.dart (hunk 5 above)
Comment (terrylucas) [reply packages/flutter_tools/test/general.shard/android/android_device_test.dart]: Limiting the parsing to just the 'App Summary' section might help with different versions.  The only differences I seen have to do with older Android version 7 I think.

Older (Android 7 I think)

`  Dalvik Heap   5110(3)    0    4136    4988(3)    0       0    9168    8958(6)  210
`

This is the modern output for a few years:
 
`  Dalvik Heap    34340    33816        0        0    62436    53883     8553
`

But I only care about the App Summary section.

**Review by terrylucas — COMMENTED** [review]
Thanks Jonah this is exactly what I needed.  Added a few suggestions and comments.

File: packages/flutter_tools/lib/src/vmservice.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/vmservice.dart]: Updated to flutterMemoryInfo

**Comment by jonahwilliams** [comment]
I've updated the example with the fields Java Heap, Native Heap, Code, Stack, Graphics, Private Other, System, and Total.

This is provided via the vmservice method `flutterMemoryInfo` (taking better name suggestions). The map also contains an identifier for the platform returned.

In the case where I fail to parse a field, I'm defaulting it to 0.

File: packages/flutter_tools/lib/src/android/android_device.dart [hunk 10]
```diff
@@ -712,6 +730,106 @@ Map<String, String> parseAdbDeviceProperties(String str) {
...
+/// For more information, see https://developer.android.com/studio/command-line/dumpsys.
+@visibleForTesting
+AndroidMemoryInfo parseMeminfoDump(String input) {
+  Iterable<String> lines = input.split('\n');
+  lines = lines
+    .skipWhile((String line) => !line.contains('App Summary'))
+    .takeWhile((String line) => !line.contains('TOTAL'));
+  final AndroidMemoryInfo androidMemoryInfo = AndroidMemoryInfo();
+  for (String line in lines) {
+    if (!line.contains(':')) {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/android/android_device.dart]: To avoid mixing styles:

Maybe move to after the `takeWhile`, like `.where((String line) => line.contains(':'))`

Then you can `.forEach((String line) {...})` instead of a for loop.

**Review by zanderso — APPROVED** [review]
lgtm w/ nit

**Review by terrylucas — APPROVED** [review]
